src/Components/InfiniteScroll.tsx

```diff
@@ -103,232 +99,290 @@ const DefaultContentPlaceHolder = ({ height }: IContentPlaceHolderProps) => {
-        },
+        updateState({
+          pageHeightMap
+        });
+      }
     },
-  ),
-  withPropsOnChange(["debounceTime"], ({ debounceTime, handleScroll }) => ({
-    handleScrollDebounced: debounce(handleScroll, debounceTime),
-    handleScrollThrottled: throttle(handleScroll, debounceTime),
-  })),
-  withHandlers({
-    handleScrollPersistingEvent: ({ handleScrollDebounced, handleScrollThrottled }) => (event: any) => {
-      const clientHeight = event.target.clientHeight;
-      const scrollHeight = event.target.scrollHeight;
-      const scrollTop = event.target.scrollTop;
-
-      handleScrollThrottled(clientHeight, scrollHeight, scrollTop);
-      handleScrollDebounced(clientHeight, scrollHeight, scrollTop);
+    handleScroll: ({
+      currentPage,
+      disableVirtualScrolling,
+      estimatedPageHeight,
+      hasNextPage,
+      lastScrollTop,
+      nextPageQuery,
+      pageHeightMap,
+      query,
+      totalPages,
+      updateState,
+    }) => ({ target: { clientHeight, scrollHeight, scrollTop } }) => () => {
+      const updatedState = {
+        lastScrollTop: scrollTop
+      } as Partial<IInfiniteScrollStateProps<any>>;
+
+      const scrollingDown = scrollTop > lastScrollTop;
+      const currentScrollBottom = scrollHeight - scrollTop - clientHeight;
+
+      if (!disableVirtualScrolling) {
+        let nextPageToLoad = 1;
+        let nextPageScrollTop = pageHeightMap[1];
+
+        while (nextPageScrollTop < (scrollTop + (clientHeight / 2)) && nextPageToLoad < totalPages) {
+          nextPageToLoad++;
+          nextPageScrollTop += nextPageToLoad in pageHeightMap
+            ? pageHeightMap[nextPageToLoad]
+            : estimatedPageHeight;
+        }
+
+        if (currentPage !== nextPageToLoad) {
+          updatedState.query = query.query.page(nextPageToLoad);
+        }
+      } else if (hasNextPage && scrollingDown && currentScrollBottom < (clientHeight / 2)) {
+        updatedState.query = nextPageQuery;
+      }
+
+      updateState(updatedState);
     },
   }),
-  lifecycle<{ disableVirtualScrolling, handleScrollDebounced, handleScrollThrottled, query: IQueryManager<any>, updatePageHeightMap }, {}>({
+  pure,
+  lifecycle<IInfiniteScrollInternalProps<any>, {}>({
     componentDidMount() {
-      if (!this.props.disableVirtualScrolling) {
-        // Initialize page height map and estimated page height
-        this.props.updatePageHeightMap(this.props.query);
+      if (this.props.query.response) {
+        this.props.updateStateFromQuery(this.props.query);
       }
     },
-    componentWillUnmount() {
-      this.props.handleScrollDebounced.cancel();
-      this.props.handleScrollThrottled.cancel();
+    componentDidUpdate(prevProps) {
+      if (prevProps.query !== this.props.query && this.props.query.response) {
+        this.props.updateStateFromQuery(this.props.query);
+      } else if (this.props.currentItems && this.props.pageHeightMap === prevProps.pageHeightMap) {
```

Comment:
I don't fully understand why we're updating the pageHeightMap if it's the same over both states.

Reply:
Since `componentDidUpdate` is fired whenever the component updates, checking that the `pageHeightMap` from current props are the same as previous props prevents an infinite cycle of constantly re-firing to update the `pageHeightMap`. In other words - if the component updated for some other reason than the `pageHeightMap` changing, then we want to update the `pageHeightMap`.

Reply:
Gotcha!